polyply/src/gen_seq.py

```diff
@@ -0,0 +1,254 @@
+    return graph
+
+
+def _random_replace_nodes(graph, residues, weights):
+    """
+    Randomly replace resname attribute of `graph`
+    based on the names in `residues` taking into
+    account `weights` provided.
+
+    Parameters
+    ----------
+    graph: `:class:networkx.Graph`
+    residues: list[str]
+    weights: list[float]
+
+    Returns:
+    --------
+    `:class:networkx.Graph`
+    """
+    for node in graph.nodes:
+        resname = random.choices(residues, weights=weights)
+        graph.nodes[node]["resname"] = resname[0]
+
+    return graph
+
+
+class MacroString():
+    """
+    Define a (random) tree graph based on a string.
+    The graph is generated each time the `gen_graph`
+    attribute is run. For random graph they will be
+    independent.
+    """
+
+    def __init__(self, string):
+        """
+        Convert string into definitions of the
+        random tree graph.
+
+        Parameters:
+        -----------
+        string: str
+           the input string containing all definitions
+        """
+        input_params = string.split(":")
+        self.name = input_params[0]
+        self.levels = int(input_params[1])
+        self.bfact = int(input_params[2])
+        self.residues = []
+        self.weights = []
+
+        for res_prob in input_params[3].split(','):
+            name, prob = res_prob.split("-")
+            self.residues.append(name)
+            self.weights.append(float(prob))
+
+    def gen_graph(self, seed=None):
+        """
+        Generate a graph from the defnitions stored in this
+        instance.
+        """
+        graph = _branched_graph("dum", self.bfact, self.levels)
+        graph = _random_replace_nodes(graph, self.residues, self.weights)
+        return graph
+
+
+def _add_edges(graph, edges, idx, jdx):
+    """
+    Add edges to a graph using the edge format
+    'node_1,node_2', where the node keys refer
+    to the node relative to the nodes with the same
+    attribute `seqid`. The sequence ids are provided by
+    `idx` and `jdx`.
+
+    Parameters
+    ----------
+    graph: `:class:networkx.Graph`
+    edges:  list[str]
+        str has the format `node_keyA,node_keyB-node_keyC,node_keyD`
+    idx: int
+    jdx: int
+
+    Returns
+    --------
+    `:class:networkx.Graph`
+    """
+    for edge in edges.split(","):
+        node_idx, node_jdx = edge.split("-")
+        idx_nodes = find_atoms(graph, "seqid", idx)
+        jdx_nodes = find_atoms(graph, "seqid", jdx)
+
+        if len(idx_nodes) == 0:
+            msg = ("Trying to add connect between block with seqid {} and block with"
+                   "seqid {}. However, cannot find block with seqid {}.")
+            raise IOError(msg.format(idx, jdx, idx))
+        elif len(jdx_nodes) == 0:
+            msg = ("Trying to add connect between block with seqid {} and block with"
+                   "seqid {}. However, cannot find block with seqid {}.")
+            raise IOError(msg.format(idx, jdx, jdx))
```

Comment:
What happens if the molecule is screwed and there are multiple/too many nodes with seqid idx? Not very likely in the near future, so could be a simple assert.
Also, idiomatic, this should probably be `if not idx_nodes: ...; if not jdx_nodes: ...`

Reply:
There cannot be multiple/too many nodes with the same seqid idx by construction.